io-engine/src/bdev/nexus/nexus_io.rs

```diff
@@ -186,53 +253,51 @@ impl<'n> NexusBio<'n> {
         self.ctx_mut().in_flight -= 1;
 
         if status == IoCompletionStatus::Success {
-            self.ok_checked();
+            self.ctx_mut().successful += 1;
         } else {
-            // IO failure, mark the IO failed and take the child out
-            error!(
-                "{:?}: IO completion for '{}' failed: {:?}, ctx={:?}",
-                self,
-                child.device_name(),
-                status,
-                self.ctx()
-            );
             self.ctx_mut().status = IoStatus::Failed;
-            self.ctx_mut().must_fail = true;
-            self.handle_failure(child, status);
+            self.ctx_mut().failed += 1;
+
+            self.completion_error(child, status);
         }
-    }
 
-    /// Complete the IO marking at as successfully when all child IO's have been
-    /// accounted for. Failing to account for all child IO's will result in
-    /// a lockup.
-    #[inline]
-    fn ok_checked(&mut self) {
-        if self.ctx().in_flight == 0 {
-            if self.ctx().must_fail {
-                //warn!(?self, "resubmitted due to must_fail");
-                self.retry_checked();
-                //self.fail();
-            } else {
-                self.ok();
-            }
+        if self.ctx().in_flight > 0 {
+            // More child I/Os to complete, not yet ready to complete nexus I/O.
+            trace_io!("Inflight: {self:?}");
+            return;
         }
-    }
 
-    /// Complete the IO marking it as failed.
-    #[inline]
-    fn fail_checked(&mut self) {
-        if self.ctx().in_flight == 0 {
+        if self.ctx().failed == 0 {
+            // No child failures, complete nexus I/O with success.
+            trace_io!("Success: {self:?}");
+            self.ok();
+        } else if self.ctx().successful > 0 {
+            // Having some child failures, resubmit the I/O.
+            self.resubmit();
+        } else {
```

Comment:
In case of single child case if the IO fails, is it intended to fail immediately as it's happening now?

Reply:
In the case of single child I/O failure, successfull == 0